theia-docker: update image to include vscode ext

vince-fugnitto opened the pull request:
<!--
Thank you for your Pull Request. Please provide a description and review
the requirements below.

Contributors guide: https://github.com/theia-ide/theia/blob/master/CONTRIBUTING.md
-->

#### What it does
<!-- Include relevant issues and describe how they are addressed. -->

The following pull-request updates the base _JavaScript_ and _TypeScript_ `theia-docker` image to include **builtin** and external **vscode** extensions. The updates include:

- updating the `latest` and `next` images to include builtin extensions and 
remove soon-to-be-deprecated extensions.
- added the external `vscode-eslint` and `vscode-editorconfig` extensions.


#### How to test
<!-- Explain how a reviewer can reproduce a bug, test new functionality or verify performance improvements. -->

Verify that the CI successfully passes for both `theia:latest` and `theia:next`.

#### Review checklist

- [x] as an author, I have thoroughly tested my changes and carefully followed [the review guidelines](https://github.com/theia-ide/theia/blob/master/doc/pull-requests.md#requesting-a-review)

#### Reminder for reviewers

- as a reviewer, I agree to behave in accordance with [the review guidelines](https://github.com/theia-ide/theia/blob/master/doc/pull-requests.md#reviewing)

Signed-off-by: vince-fugnitto <vincent.fugnitto@ericsson.com>

akosyakov commented on theia-docker/latest.package.json:
```diff
@@ -12,32 +12,97 @@
     },
     "dependencies": {
         "@theia/callhierarchy": "latest",
+        "@theia/cli": "latest",
```
`@theia/cli` should always go to dev dependecy

vince-fugnitto replied:
I understand but what should we do in such a case where we build the `—production` flag. It means that I will not have the `download:plugins` script for such applications.

akosyakov replied:
This script should be called during the build time, not at runtime. At the build time `@theia/cli` is present.

akosyakov commented on theia-docker/latest.package.json:
```diff
@@ -12,32 +12,97 @@
-        "@theia/textmate-grammars": "latest",
-        "@theia/tslint": "latest",
-        "@theia/typescript": "latest",
         "typescript": "latest"
```
`typescript` should be removed, it is already bundled into built-in

akosyakov commented on theia-docker/latest.package.json:
```diff
@@ -12,32 +12,97 @@
+        "vscode-builtin-lua": "https://github.com/theia-ide/vscode-builtin-extensions/releases/download/v1.39.1-prel/lua-1.39.1-prel.vsix",
+        "vscode-builtin-make": "https://github.com/theia-ide/vscode-builtin-extensions/releases/download/v1.39.1-prel/make-1.39.1-prel.vsix",
+        "vscode-builtin-markdown": "https://github.com/theia-ide/vscode-builtin-extensions/releases/download/v1.39.1-prel/markdown-1.39.1-prel.vsix",
+        "vscode-builtin-markdown-language-features": "https://github.com/theia-ide/vscode-builtin-extensions/releases/download/v1.39.1-prel/markdown-language-features-1.39.1-prel.vsix",
```
let's remove it and stick with `@theia/preview` for now

akosyakov commented on theia-docker/latest.package.json:
```diff
@@ -12,32 +12,97 @@
+        "@theia/cli": "latest",
         "@theia/debug": "latest",
-        "@theia/debug-nodejs": "latest",
+        "@theia/editor-preview": "latest",
```
please remove, i don't like how it works and did not include on purporse

akosyakov commented:
please have a look at comments and don't forget to fix them in next image as well

akosyakov commented on theia-docker/Dockerfile:
```diff
@@ -13,7 +13,8 @@ RUN yarn --pure-lockfile && \
     yarn autoclean --force && \
-    yarn cache clean
+    yarn cache clean ; \
+    yarn theia download:plugins
```
you should call it before `yarn --production`.

Did me merge `@theia/cli` as a runtime dependency to some images already? we have to undo it

akosyakov commented:
@marcdumais-work @vince-fugnitto I wonder whether we can add at least a flag to download:plugins script that it unarchive extensions. It speeds up the start like 10x and does not pollute the disk by unzipping at runtime.

vince-fugnitto commented:
> @marcdumais-work @vince-fugnitto I wonder whether we can add at least a flag to download:plugins script that it unarchive extensions. It speeds up the start like 10x and does not pollute the disk by unzipping at runtime.

https://github.com/eclipse-theia/theia/pull/7123

akosyakov commented:
Could someone test please?  It looks great, but i hardly can find time to try.

marcdumais-work commented:
> > @marcdumais-work @vince-fugnitto I wonder whether we can add at least a flag to download:plugins script that it unarchive extensions. It speeds up the start like 10x and does not pollute the disk by unzipping at runtime.
> 
> [eclipse-theia/theia#7123](https://github.com/eclipse-theia/theia/pull/7123)

So we should benefit from this commit immediately for the `next` version of the image, and `latest` after end-of-month release. So it's expected that `latest` will have a slower startup for a couple more weeks.

vince-fugnitto commented:
> > > @marcdumais-work @vince-fugnitto I wonder whether we can add at least a flag to download:plugins script that it unarchive extensions. It speeds up the start like 10x and does not pollute the disk by unzipping at runtime.
> > 
> > 
> > [eclipse-theia/theia#7123](https://github.com/eclipse-theia/theia/pull/7123)
> 
> So we should benefit from this commit immediately for the `next` version of the image, and `latest` after end-of-month release. So it's expected that `latest` will have a slower startup for a couple more weeks.

Yes, since the flag was added in `next`, the `latest` image will be a bit slower on startup since the plugins are preserved as `.vsix`. Is it a blocker? (in my opinion I think it's fine to be a bit slower on startup until the next release).

marcdumais-work commented on theia-docker/Dockerfile:
```diff
@@ -28,7 +29,8 @@ ENV HOME /home/theia
 EXPOSE 3000
-ENV SHELL /bin/bash
+ENV SHELL=/bin/bash \
+    THEIA_DEFAULT_PLUGINS=local-dir:/home/theia/plugins
```
Should we document somewhere that this is the folder where a user can add VS Code extensions on their own? In application-level README?

vince-fugnitto replied:
> Should we document somewhere that this is the folder where a user can add VS Code extensions on their own? In application-level README?

Since all images will eventually be updated to include plugin support (with builtins), should we document it globally (root of the repo), or per application?

marcdumais-work replied:
However I am not sure we'll update all the images in the repo to use plugins. Or at least some will be low-priority and probably be "PR welcome" situation.

vince-fugnitto replied:
> However I am not sure we'll update all the images in the repo to use plugins. Or at least some will be low-priority and probably be "PR welcome" situation.

@marcdumais-work I think all images at some point will need to be updated to make use of the builtins and remove deprecated extensions, but you are right that it can be done by other members of the community. Should I update the `readme` for each image I touch? I think some images themselves do not have readmes so they should align. I can also do a pass later if you prefer and update the readmes of the images and add new ones if none exist.

marcdumais-work replied:
I think it will be ideal to have a README for each image. But for now you can add this info in the main README.

vince-fugnitto replied:
> I think it will be ideal to have a README for each image. But for now you can add this info in the main README.

I'll do the updates as part of https://github.com/theia-ide/theia-apps/pull/279

lmcbout commented:
I created the next and latest docker images . Both compule and was able to run it.
I think it would be good to say how to "Build" and how to "RUN" in a README file

vince-fugnitto commented:
> I created the next and latest docker images . Both compule and was able to run it.
> I think it would be good to say how to "Build" and how to "RUN" in a README file

@lmcbout is there anything left to do before it can be approved?

Updates to the `README` will be done at a later step as discussed https://github.com/theia-ide/theia-apps/pull/300#discussion_r378967169, and for now you can see build and running steps in the [contributing.md](https://github.com/theia-ide/theia-apps/blob/master/CONTRIBUTING.md).

marcdumais-work commented:
@vince-fugnitto is `eslint` working for you?

marcdumais-work commented:
> @vince-fugnitto is `eslint` working for you?

I cloned the Theia repo in the container to test a bit. Looks like I get `eslint` errors in the editor for the `latest` version of the image in this PR but not for `next`. 

Outside docker `eslint` works for me when I build the browser example from master, which should be close to the `next` used in this image. Any ideas?

marcdumais-work commented:
I see exceptions in the backend log. However I see the same or very similar with the example app in the main repo built from sources, so probably not a problem specific to the image. 

```
root ERROR [hosted-plugin: 202] Error: Possible Emitter memory leak detected. 31 listeners added. Use event.maxListeners to increase the limit (30)
    at Emitter.checkMaxListeners (/home/theia/node_modules/@theia/core/lib/common/event.js:253:26)
    at PreferenceRegistryExtImpl._event.Object.assign.maxListeners [as onDidChangeConfiguration] (/home/theia/node_modules/@theia/core/lib/common/event.js:221:27)
    at Object.onDidChangeConfiguration (/home/theia/node_modules/@theia/plugin-ext/lib/plugin/plugin-context.js:384:46)
    at b.start (/home/theia/plugins/vscode-builtin-gulp/extension/dist/main.js:1:4339)
    at t.activate (/home/theia/plugins/vscode-builtin-gulp/extension/dist/main.js:1:6103)
    at PluginManagerExtImpl.<anonymous> (/home/theia/node_modules/@theia/plugin-ext/lib/plugin/plugin-manager.js:469:87)
    at step (/home/theia/node_modules/@theia/plugin-ext/lib/plugin/plugin-manager.js:47:23)
    at Object.next (/home/theia/node_modules/@theia/plugin-ext/lib/plugin/plugin-manager.js:28:53)
    at fulfilled (/home/theia/node_modules/@theia/plugin-ext/lib/plugin/plugin-manager.js:19:58)
root INFO [hosted-plugin: 202] PLUGIN_HOST(202): PluginManagerExtImpl/loadPlugin(/home/theia/plugins/vscode-builtin-grunt/extension/dist/main)
root ERROR [hosted-plugin: 202] Error: Possible Emitter memory leak detected. 32 listeners added. Use event.maxListeners to increase the limit (30)
    at Emitter.checkMaxListeners (/home/theia/node_modules/@theia/core/lib/common/event.js:253:26)
    at PreferenceRegistryExtImpl._event.Object.assign.maxListeners [as onDidChangeConfiguration] (/home/theia/node_modules/@theia/core/lib/common/event.js:221:27)
```

vince-fugnitto commented:
> @vince-fugnitto is `eslint` working for you?

`eslint` is successfully working for me:

<div align='center'>

<img width="1291" alt="Screen Shot 2020-02-17 at 3 59 03 PM" src="https://user-images.githubusercontent.com/40359487/74686130-866a0880-519e-11ea-98cf-ff700042c043.png">

</div>

I created a small example project:

```bash
mkdir eslint-tmp
cd eslint-tmp
npm init -y
npm install eslint --save-dev
npm install typescript --save-dev
npx eslint --init # perform the necessary steps
npm install
touch a.ts
```

I updated the rules to include:

```json
"rules": {
  "semi": ["error", "always"],
  "quotes": ["error", "single"]
}
```

And added a violation in `a.ts`

```ts
console.log("test")
```

marcdumais-work commented:
> > @vince-fugnitto is `eslint` working for you?
> 
> `eslint` is successfully working for me:

Can you confirm you tested this with `next`?

vince-fugnitto commented:
> > > @vince-fugnitto is `eslint` working for you?
> > 
> > 
> > `eslint` is successfully working for me:
> 
> Can you confirm you tested this with `next`?

I tested the following example for `eslint` using both `latest` and `next`.

marcdumais-work commented:
> I tested the following example for `eslint` using both `latest` and `next`.

I re-tried, using the Theia main repo as test case, and `eslint` now works OOTB.

vince-fugnitto commented:
> LGTM, thanks @vince-fugnitto

Thank you for the review! :)
